Support configurable hostPort in helm chart

Gasoid opened the pull request:
### Proposed changes

hostPort field for nginx container is configurable

Problem: hostPort is not configurable for now

Solution: In values file were added hostPort section

Testing: Tested manually.

Closes #2933 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginx/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### Release notes

```release-note
NONE
```

Gasoid commented:
I have hereby read the F5 CLA and agree to its terms

bjee19 commented:
Hey @Gasoid thanks for opening up this PR! We're going to hold off on this change right now because our architecture is changing in our next release (2.0) and there will be a different way to specify fields on the nginx container.

We are about to merge our feature branch `change/control-data-plane-split` into `main`, hopefully within the next week, and when that happens these changes would no longer work. 

So after we merge our feature branch you could rewrite this PR to fit the new changes, it would probably look something like this PR https://github.com/nginx/nginx-gateway-fabric/pull/3319, OR we can probably just tag it alongside this issue https://github.com/nginx/nginx-gateway-fabric/issues/3240 which we'll complete by our 2.0 release. 

We're fine either way, so please let us know what you'd like to do!

Gasoid commented:
@bjee19 thx, i will take care of #3240 
according to you, all issues with label "good first issue" are not needed to start doing?
e.g. i would like to do #3157 as well

bjee19 commented:
Hey @Gasoid, thanks for offering to do #3240, however that won't be necessary. That specific story is already scheduled in our [current sprint](https://github.com/orgs/nginx/projects/10/views/4) so one of our team members will already be handling that. However, modifying this open PR after we merge our control data plane split changes into main is still an option and would be greatly appreciated to close out #2933. 

>according to you, all issues with label "good first issue" are not needed to start doing?

That sort of depends, we're in a tricky spot right now where we're just wrapping up our changes to release our 2.0 with major changes to our architecture, so it just means that the solution to some of these stories might change slightly. However after we merge our changes into main in this [open PR](https://github.com/nginx/nginx-gateway-fabric/pull/3318) (should be by next week), all these issues are good to go. 

>e.g. i would like to do https://github.com/nginx/nginx-gateway-fabric/issues/3157 as well

#3157 looks like a great issue that won't be affected too much with the changes coming in so you could start on that. There just might be some additional places to update after #3318 goes in.

Gasoid commented:
> However, modifying this open PR after we merge our control data plane split changes into main is still an option and would be greatly appreciated to close out https://github.com/nginx/nginx-gateway-fabric/issues/2933
ok, i will modify this PR once you have done everything with release

bjee19 commented:
Hey @Gasoid, we've merged our `change/control-data-plane-split` into `main`, thus completing our big restructure change to our architecture. So if you'd still like to revise this PR feel free to, and please let us know if you have any questions!

Gasoid commented:
yes, i am going to revise the PR next week

sjberman commented on charts/nginx-gateway-fabric/templates/deployment.yaml:
```diff
@@ -135,6 +135,9 @@ spec:
         ports:
         - name: agent-grpc
           containerPort: 8443
+          {{- if .Values.nginx.hostPort.enable }}
+          hostPort: {{ .Values.nginx.hostPort.port }}
+          {{- end }}
```
This deployment no longer for nginx, it's just the control plane.

The nginx deployment is built in-code by the control plane. https://github.com/nginx/nginx-gateway-fabric/pull/3319 is an example of how we now add a new field to the nginx deployment/service, so you'll need to do something similar to add these new hostPort fields. The fields need to exist in the NginxProxy CRD, which is populated at installation time via the helm chart (or at runtime via the k8s API), and is then ingested by the control plane to configure and deploy the nginx deployment.

Gasoid replied:
ok, i will adjust nginxProxy CRD in order to support this hostPort field

sjberman commented on internal/controller/provisioner/objects.go:
```diff
@@ -780,6 +780,11 @@ func (p *NginxProvisioner) buildNginxPodTemplateSpec(
 				container.Args = append(container.Args, "debug")
 			}
+
+			if containerSpec.HostPort != nil && containerSpec.HostPort.Enable && len(container.Ports) > 0 {
+				container.Ports[0].HostPort = containerSpec.HostPort.Port
+			}
```
The issue with this approach is that the nginx container is going to be exposed on X number of ports, which are directly correlated to the listener ports in the Gateway object.

You've only exposed a single hostPort, and are binding it to whatever the first container port is. When in reality, each container port should have its own hostPort. If you take a look at how we added NodePorts in the `values.yaml` file and then how we use that to set Service ports; it should be a similar structure, where each port maps to a Gateway listener port.

sjberman commented on apis/v1alpha2/nginxproxy_types.go:
```diff
@@ -608,3 +613,16 @@ type NodePort struct {
+
+// HostPort to expose on the host.
+type HostPort struct {
+	// ContainerPort is nginx port.
```
```suggestion
	// ContainerPort is the port on the nginx container to map to the HostPort.
```

sjberman commented on apis/v1alpha2/nginxproxy_types.go:
```diff
@@ -608,3 +613,16 @@ type NodePort struct {
 	ListenerPort int32 `json:"listenerPort"`
 }
+
+// HostPort to expose on the host.
```
```suggestion
// HostPort exposes an nginx container port on the host.
```

Gasoid commented:
@sjberman i have adjusted code according to your comments

sjberman commented on charts/nginx-gateway-fabric/templates/nginxproxy.yaml:
```diff
@@ -19,7 +19,9 @@ spec:
       {{- end }}
       container:
         {{- if .Values.nginx.container }}
```
I don't think you should need this `if` statement (or below) if you're using `with`.

sjberman commented on charts/nginx-gateway-fabric/templates/_helpers.tpl:
```diff
@@ -102,5 +102,7 @@ Filters out empty fields from a struct.
   {{- end }}
 {{- end }}
-{{- $result | toYaml }}
+{{- if $result }}
+  {{- $result | toYaml }}
+{{- end }}
```
Can you explain this a bit more? Not sure why this is needed?

Gasoid replied:
that was issue with empty result and toYaml. I got yaml issue
since i removed `if .Values.nginx.container` the issue has gone.

Gasoid replied:
if result is empty, toYaml fails.
default `container` section:

```yaml
container:
  hostPorts: []
```

so $result will be empty/null, because hostPorts is empty.

`service` section works, because it had non-empty keys.

Gasoid replied:
similar issue: https://github.com/helm/helm/issues/5261

Gasoid commented on charts/nginx-gateway-fabric/templates/_helpers.tpl:
```diff
@@ -102,5 +102,4 @@ Filters out empty fields from a struct.
     {{- $result = merge $result (dict $key $value) }}
   {{- end }}
 {{- end }}
-{{- $result | toYaml }}
```
no, typo, tested without if,
it doesn't work if container block is empty

sjberman commented on charts/nginx-gateway-fabric/values.yaml:
```diff
@@ -399,7 +399,34 @@ nginx:
+    # Use this option when you need to expose container ports on the host for direct access,
+    # such as for debugging, legacy integrations, or when NodePort/LoadBalancer services are not suitable.
+    # Note: Using hostPort may have security and scheduling implications, as it ties pods to specific nodes and ports.
+    hostPorts: []
```
Yeah fixing that empty newline would be nice. I would think adding another `-` to the template lines, although that may result in config being pulled up to the wrong line.

Gasoid replied:
solved

sjberman commented:
Something is failing with the installation or was not generated correctly in the CRD yaml:

```
Error: INSTALLATION FAILED: failed to install CRD crds/gateway.nginx.org_nginxproxies.yaml: 1 error occurred:
	* CustomResourceDefinition.apiextensions.k8s.io "nginxproxies.gateway.nginx.org" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[kubernetes].properties[daemonSet].properties[container].properties[hostPorts].items.type: Required value: must not be empty for specified array items
```

Gasoid commented:
I found another issue, I will try to fix it this week later

Пн, 16 июня 2025 г. в 22:50, Saylor Berman ***@***.***>:

> *sjberman* left a comment (nginx/nginx-gateway-fabric#3321)
> <https://github.com/nginx/nginx-gateway-fabric/pull/3321#issuecomment-2978090899>
>
> Something is failing with the installation or was not generated correctly
> in the CRD yaml:
>
> Error: INSTALLATION FAILED: failed to install CRD crds/gateway.nginx.org_nginxproxies.yaml: 1 error occurred:
> 	* CustomResourceDefinition.apiextensions.k8s.io "nginxproxies.gateway.nginx.org" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[kubernetes].properties[daemonSet].properties[container].properties[hostPorts].items.type: Required value: must not be empty for specified array items
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nginx/nginx-gateway-fabric/pull/3321#issuecomment-2978090899>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGLNBLVQM2SOVS7YSSSKQT3D4U2BAVCNFSM6AAAAAB3ZXVEWOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNZYGA4TAOBZHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Gasoid commented:
> Something is failing with the installation or was not generated correctly in the CRD yaml:
> 
> ```
> Error: INSTALLATION FAILED: failed to install CRD crds/gateway.nginx.org_nginxproxies.yaml: 1 error occurred:
> 	* CustomResourceDefinition.apiextensions.k8s.io "nginxproxies.gateway.nginx.org" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[kubernetes].properties[daemonSet].properties[container].properties[hostPorts].items.type: Required value: must not be empty for specified array items
> ```

i fixed this failure and adjusted another place where it could appear. CRDs were tested

bjee19 commented:
great lgtm, thanks!

sjberman commented on charts/nginx-gateway-fabric/values.yaml:
```diff
@@ -402,7 +402,33 @@ nginx:
+    #   containerPort: 80
+
     # -- The resource requirements of the NGINX container.
     # resources: {}
```
Since you updated the template to filter empty fields, can we uncomment the other fields in the `container` resource to have them render in the schema/README, without them being rendered in the CRD if empty?

Gasoid replied:
ok, let me do this

sjberman commented on apis/v1alpha2/nginxproxy_types.go:
```diff
@@ -608,3 +613,16 @@ type NodePort struct {
+// HostPort exposes an nginx container port on the host.
+type HostPort struct {
+	// Port to expose on the host.
+	// kubebuilder:validation:Minimum=1
+	// kubebuilder:validation:Maximum=65535
```
This actually needs to have a `+` before `kubebuilder` in order to render properly. It looks like our NodePort struct has this mistake as well. Do you mind updating all of these?

`// +kubebuilder:validation:Minimum=1` for example.

Gasoid commented:
Thx you @sjberman  for support
